addons/payment_paypal/models/paypal_s2s.py

```diff
@@ -0,0 +1,231 @@
+                    'credit_card_token': {
+                        'credit_card_id': str(self.payment_method_id.acquirer_ref),
+                        'payer_id': str(self.payment_method_id.partner_id.id)
+                    }
+                }]
+            },
+            'transactions': [{
+                'amount': {
+                    'total': str(amount),
+                    'currency': str(self.currency_id.name),
+                },
+                'description': str(self.display_name)
+            }]
+        }
+        payment_result = dict()
+        try:
+            payment_request = urllib2.Request(payment_url,
+                                              json.dumps(data),
+                                              payment_headers)
+            payment_result = self._paypal_try_url(payment_request, tries=3)
+
+        except Exception, e:
+            _logger.exception('PayPal create payment request failed - %s' % e)
+            raise UserWarning(_('PayPal create payment request failed.'))
+        return payment_result
+
+    @api.multi
+    def paypal_s2s_do_transaction(self, **vals):
+        self.ensure_one()
+        payment_method = self.env['payment.method'].search([
+            ('partner_id', '=', self.partner_id.id),
+            ('acquirer_id', '=', self.acquirer_id.id)])
+        if not payment_method:
+            raise UserWarning(_('Credit/Debit card has been not added in your account.'))
+
+        payment_result = self._paypal_s2s_send()
+
+        return self._paypal_s2s_validate_tree(payment_result)
+
+    @api.cr_uid_records_context
+    def paypal_s2s_form_validate(self, cr, uid, tx, context=None):
+        tree = self._paypal_s2s_get_tx_status(tx)
+        return self._paypal_s2s_validate_tree(tx, tree)
+
+    @api.multi
+    def _paypal_s2s_validate_tree(self, data):
+        response = data['state']
+        if response == 'approved':
+            state_msg = 'Received response for Paypal payment tx %s: '\
+                        'set as %s' % (self.reference, response)
+            _logger.info(state_msg)
+            self.write({
+                'state': 'done',
+                'acquirer_reference': data['id'],
+                'state_message': _(state_msg),
+                'date_validate': data['update_time']
+            })
+            if self.callback_eval:
+                safe_eval(self.callback_eval, {'self': self})
+            return True
+        elif response in ['pending', 'expired', 'failed', 'canceled']:
+            response = 'cancel' if response == 'canceled' else response
+            state_msg = 'Received response for Paypal payment '\
+                        'transaction %s: set as %s' % (self.reference, response)
+            _logger.info(state_msg)
+            self.write({
+                'state': response,
+                'acquirer_reference': data['id'],
+                'state_message': _(state_msg),
+            })
+            return True
+        else:
```

Comment:
If the state is pending, you should `time.sleep()` to wait a bit then check the status again (see ogone implementation)